### PR TITLE
Bug 789 - Create a save mission state that includes position, orientation and heading 

### LINKED ISSUE #789: Create startup states and save mission state
> This seems so obvious, but I haven't found a way to do this in fg 2016.2.1. I want to fly to an airport and stay there "overnight" and continue my trip the next time I start flightgear. The plain should be where I parked it on that airport I landed on.
> I really found no way to do that and I searched online for quite some time to find an answer.
> Thanks,
> Jay

See https://forum.flightgear.org/viewtopic.php?f=17&t=29993#p290183

Think it would be possible to implement something like this as a saved state to allow a convenient way to continue a session?

Other thought's would be adapting the flight recorder to accomplish the same thing?

Just a thought, we may want to discuss it.

### PR DESCRIPTION
**Pull request by wlbragg**
This is so simple to do. I can't believe it took this long for us to attempt to implement. It should fit nicely with the "save state of aircraft between sessions".

I am also implementing this on the AirCrane and the J3Cub.

Fixes #789

### CODE REVIEW

**Comment by wlbragg** [comment]
This is probably not ready to merge. I think it needs to be thoroughly tested first. I added it to the Cub which should help expose any unforeseen bugs over time of which I can respond to quickly if need be.
I'm currently trying to test improper game exits, game crashes, etc. to find any potential unforeseen issues.

**Comment by gilbertohasnofb** [comment]
> This is probably not ready to merge. I think it needs to be thoroughly tested first.

How exactly should this be tested?

**Comment by wlbragg** [comment]
I guess just use it and see if you can break it. I added some conditions to the saving of the location so it only saves your position if you are on the ground and close to stopped (and of course have the "Persistent position" check-box checked".
I did this so it will not save if your crash in flight or landing somewhere. Also it shouldn't compete with the saved or startup states work done prior. Like if you are using a "startup state" like approach and also have the "Persistent position" checked, it should ignore the persistent position and still start up on approach. I have actually tested that yet, but it should work that way.
I considered having to click a "Save" button each time you wanted to save it at a location but opted for this simpler approach.
I think there are a few ways you could do this...

1) the way it is now, if "Persistent position" is checked it will save and load that position if all conditions are meant.

2) you could have a save and load dialog requiring an action to both save or load a position.

3) you could have both methods with the "Persistent position" checked overriding the save/load GUI.

4) you could have a save and load dialog requiring an action to both save or load a position. It could have multiple save/load positions that you could choose from.

5) you could have a check for a saved position on startup and ask if you want to activate it.

I opted for the way it is because it was simple and effective with a low overhead.

**Comment by wlbragg** [comment]
>Like if you are using a "startup state" like approach and also have the "Persistent position" checked, it should ignore the persistent position and still start up on approach. 

Now that I think about it, it may not ignore "Persistent position" when starting a save state that starts the aircraft stopped and on the ground. I'll test that now.

**Comment by wlbragg** [comment]
This isn't playing nicely with GUI-Location/Select Airport. It needs a check for this condition/action if possible.

**Comment by gilbertohasnofb** [comment]
@wlbragg sorry for my late reply. I will try to test this and report back here as soon as I can, but I will be very busy during these next weeks and I might not have a chance to do that soon.

**Comment by wlbragg** [comment]
I can't figure out a way to catch the GUI-Location/Select Airport action so as to stop the repositioning on load if "Persistent Position" is checked. (GUI-Location/Select Airport uses an fgcommand("reposition") to move the aircraft to a new airport location. I may not be able to "automatically" have the aircraft start at the saved or "persistent" location. Instead we may have to have a Save/Load feature.

**Comment by wlbragg** [comment]
I'm going to talk about this on the dev list and see if we don't want to make this a system wide feature in the Location Dialog.

**Comment by onox** [comment]
I like the feature, but I managed to break it when I tried to spawn at a different airport (stored at KLSV, loaded at FHAW).

File: Nasal/c172p.nas [hunk 1]
```diff
@@ -393,6 +393,17 @@ var reset_system = func {
     props.globals.getNode("/fdm/jsbsim/pontoon-damage/right-pontoon", 0).setIntValue(0);
 
     setprop("/engines/active-engine/kill-engine", 0);
+
+    if (getprop("/fdm/jsbsim/position/persistent")) {
+        setprop("position/latitude-deg", getprop("/sim/model/c172p/currentlat"));
```
Comment (onox) [inline Nasal/c172p.nas]: You have a lot of properties in the same directory that are related to each other (`/sim/model/c172p/current*`). Maybe you could group them together and move them to a separate directory like `/sim/model/c172p/current-position/*`)?

**Comment by onox** [comment]
Maybe store the closest airport and check if the current airport (when spawning) is the same as the stored one?

**Comment by onox** [comment]
(or measure the distance between the spawned position and the stored position)

**Comment by wlbragg** [comment]
> I managed to break it when I tried to spawn at a different airport (stored at KLSV, loaded at FHAW

I have found it can break with two other actions.
1) Go to airport using GUI
2) Saving the position over water. I can't figure out why when saved over water it always drops the aircraft from about 10-20 feet and crashes, even though I save the altitude or make it 0.
I'll try breaking it the way you did also and see what that is all about.

>Maybe you could group them together and move them to a separate directory 
Will do.

I talked to James on the mailing list and he would like this done using the reposition cmd/code. Potentially that might fix all the bugs it has now. I'll keep working that angle and see where it goes.

**Comment by wlbragg** [comment]
Thanks for looking at it and the comments.

**Comment by wlbragg** [comment]
This needs to be put on hold for now and put to an issue for work on at a later date.

**Comment by gilbertohasnofb** [comment]
@wlbragg sorry I have not tested this yet, I have been just so busy lately... once you solve this issue and make a new PR I will be happy to test it!

**Pull request closed without merging**